lib/blockchain.py

```diff
@@ -247,83 +394,175 @@ def read_header(self, height):
+        else:
+            if headers is None:
+                headers = {}
+
+            pow_limit = NetworkConstants.POW_LIMIT
+            height -= 1
+            last = headers[height] if height in headers else self.read_header(height)
+
+            if last is not None:
+                first = last
+                total = 0
+                i = 0
+
+                while i < NetworkConstants.POW_AVERAGING_WINDOW and first is not None:
+                    total += self.bits_to_target(int(first.get('bits'), 16))
+                    prev_height = height - i - 1
+                    first = headers[prev_height] if prev_height in headers else self.read_header(prev_height)
+                    i += 1
+
+                if first is None:
+                    new_target = pow_limit
+                else:
+                    avg = total // NetworkConstants.POW_AVERAGING_WINDOW
+                    actual_timespan = self.get_mediantime_past(headers,
+                                                               last.get('block_height')) - self.get_mediantime_past(
+                        headers, first.get('block_height'))
+
+                    if actual_timespan < min_actual_timespan():
+                        actual_timespan = min_actual_timespan()
+
+                    if actual_timespan > max_actual_timespan():
+                        actual_timespan = max_actual_timespan()
+
+                    avg = avg // averaging_window_timespan()
+                    avg *= actual_timespan
+
+                    if avg > pow_limit:
+                        avg = pow_limit
+
+                    new_target = int(avg)
+            else:
+                new_target = pow_limit
+
         return new_target
 
+    def get_mediantime_past(self, headers, start_height):
+        header = headers[start_height] if start_height in headers else self.read_header(start_height)
+
+        times = []
+        i = 0
+
+        while i < 11 and header is not None:
+            times.append(header.get('timestamp'))
+            prev_height = start_height - i - 1
+            header = headers[prev_height] if prev_height in headers else self.read_header(prev_height)
+            i += 1
+
+        times.sort()
+        return times[(len(times) // 2)]
+
     def bits_to_target(self, bits):
-        bitsN = (bits >> 24) & 0xff
-        if not (bitsN >= 0x03 and bitsN <= 0x1d):
-            raise BaseException("First part of bits should be in [0x03, 0x1d]")
-        bitsBase = bits & 0xffffff
-        if not (bitsBase >= 0x8000 and bitsBase <= 0x7fffff):
-            raise BaseException("Second part of bits should be in [0x8000, 0x7fffff]")
-        return bitsBase << (8 * (bitsN-3))
+        size = bits >> 24
+        word = bits & 0x007fffff
+
+        if size <= 3:
+            word >>= 8 * (3 - size)
+            ret = word
+        else:
+            ret = word
+            ret <<= 8 * (size - 3)
+
+        return ret
 
     def target_to_bits(self, target):
```

Comment:
Maybe it's better to handle it in the standard way. This is my Python implementation of the original C++ code:

```python
def uint256_to_compact(u):
    assert u >= 0
    nsize = (bits(u) + 7) // 8
    if nsize <= 3:
        c = u << (8 * (3 - nsize))
    else:
        c = u >> (8 * (nsize - 3))
    if c & 0x00800000:
        c >>= 8
        nsize += 1
    assert (c & ~0x007fffff) == 0
    assert nsize < 256
    c |= nsize << 24
    return c
```

Reply:
Sure, no problem.

Reply:
Done